### PR TITLE
Add saved_model support for tensorflow model

### PR DESCRIPTION
**Pull request by austinzh**
Currently MLeap Tensorflow Bundle use frozen graph as model format.
But frozen graph do not support ResourceVariable which is the default variable create by `tf.keras.layers.Embedding`
And starting from tensorflow 2.0 saved_model format become the official tensorflow serialization format for both keras model and tensorflow model.

It's will be great to add saved_model as a addition bundle format inside tensorflow bundle.

This PR is inspired by https://github.com/spotify/zoltar/blob/b2c4c86f06c043aae505c533467e8a42d12da2d8/zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/TensorFlowModel.java

And to keep the interface as same as graph mode, we keep using the old session.run() API instead of [ConcreteFunction API](https://github.com/tensorflow/java/blob/026863409522010869f1c4a9c8e611d217c96c41/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ConcreteFunction.java#L40)

This change is backward compatible

### CODE REVIEW

**Comment by austinzh** [comment]
@ancasarb @jsleight @voganrc 
As described, add saved_model support for tensorflow.
Feel free to review it when you  have time.

File: mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowModel.scala [hunk 1]
```diff
@@ -64,26 +70,34 @@ case class TensorflowModel(@transient var graph: Option[tensorflow.Graph] = None
...
 
-    val s = session match {
-      case Some(sess) => sess
-      case _ => { // can also be null at deserialization time, not just empty
-        session = Some(new tensorflow.Session(g))
-        session.get
-      }
-    }
+  private def getSessionFromFrozenGraph: (tensorflow.Session, tensorflow.Graph) = {
+    val graph = new tensorflow.Graph()
```
Comment (jsleight) [inline mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowModel.scala:87]: does this have a namespace collision with the class `graph` attribute?  I think no?  But might be better to pick another name?

File: mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala [hunk 2]
```diff
@@ -71,15 +71,18 @@ class TensorflowTransformerOp extends MleapOp[TensorflowTransformer, TensorflowM
...
       })
-
-      val graph = new org.tensorflow.Graph()
-      val graphDef = GraphDef.parseFrom(graphBytes)
-      graph.importGraphDef(graphDef)
-      TensorflowModel(graph = Some(graph),
+      val modelBytes = format match {
+        case Some("graph") | None => Files.readAllBytes(context.file("graph.pb"))
+        case Some("saved_model") => Files.readAllBytes(context.file("saved_model.zip"))
+        case _ => throw new RuntimeException("Only support `saved_model` and `tf` format");
```
Comment (jsleight) [inline mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala:77]: should this exception string be `saved_model and graph format`?

and maybe`UnsupportedOperationException` to be a bit more specific? (maybe a few other places too)

File: mleap-tensorflow/README.md [hunk 3]
```diff
@@ -1,34 +1,18 @@
 # Tensorflow
-
```
Comment (jsleight) [inline mleap-tensorflow/README.md:2]: there is a RELEASE_NOTES.md file in the repo root as well.  Can you add a line in there about the change.

**Review by jsleight — COMMENTED** [review]
few nitpick things, but otherwise lgtm!

File: mleap-tensorflow/README.md (hunk 3 above)
Comment (austinzh) [reply mleap-tensorflow/README.md:2]: Sure.

File: mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowModel.scala (hunk 1 above)
Comment (austinzh) [reply mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowModel.scala:87]: will do

File: mleap-tensorflow/README.md (hunk 3 above)
Comment (austinzh) [reply mleap-tensorflow/README.md:2]: updated.

File: RELEASE_NOTES.md [hunk 4]
```diff
@@ -3,12 +3,12 @@
 ### Breaking Changes
 
 ### New Features
-
 ### Bug Fixes
 - Fix OneHotEncoder Python serialization
 
 ### Improvements
 - Upgrade to scikit-learn 0.22
+- Support for tensorflow saved_model format
```
Comment (jsleight) [inline RELEASE_NOTES.md:11]: looks like it needs to resolve conflicts with master branch

**Review by jsleight — APPROVED** [review]
I'm also not sure why the travis tests didn't run, but assuming those work, then this lgtm

File: mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala [hunk 5]
```diff
@@ -71,15 +71,18 @@ class TensorflowTransformerOp extends MleapOp[TensorflowTransformer, TensorflowM
...
       })
-
-      val graph = new org.tensorflow.Graph()
-      val graphDef = GraphDef.parseFrom(graphBytes)
-      graph.importGraphDef(graphDef)
-      TensorflowModel(graph = Some(graph),
+      val modelBytes = format match {
+        case Some("graph") | None => Files.readAllBytes(context.file("graph.pb"))
+        case Some("saved_model") => Files.readAllBytes(context.file("saved_model.zip"))
+        case _ => throw new UnsupportedOperationException("Only support `saved_model` and `graph` format")
```
Comment (ancasarb) [inline mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala:77]: I don't think this is backward compatible, as bundles serialized before won't have the format set, so in that case, we should just assume `graph`.

**Review by ancasarb — CHANGES_REQUESTED** [review]
looks good overall, one change is needed (i think) to make this backward compatible.

File: mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala (hunk 5 above)
Comment (voganrc) [reply mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala:77]: I think this should be covered by the ` | None` part of the first case

File: mleap-tensorflow/src/test/scala/ml/combust/mleap/tensorflow/TensorflowModelSpec.scala [hunk 6]
```diff
@@ -44,4 +54,51 @@ class TensorflowModelSpec extends FunSpec {
       }
     }
   }
+
+  describe("with an tensorflow model has variables") {
+    it("saved model") {
+      var reducedSum = 0.0f
+      val testFolder = Files.createTempDirectory("tf-saved-model-export")
```
Comment (voganrc) [inline mleap-tensorflow/src/test/scala/ml/combust/mleap/tensorflow/TensorflowModelSpec.scala:61]: In `TensorFlowModel.scala` you manually call `FileUtil().rmRF` on your temp directory, so I just want to make sure that cleanup is not necessary in this test

**Review by voganrc — APPROVED** [review]
LGTM!

File: mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala (hunk 5 above)
Comment (ancasarb) [reply mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowTransformerOp.scala:77]: ah yes, sorry i didn't see the or part of that case.

File: mleap-tensorflow/src/test/scala/ml/combust/mleap/tensorflow/TensorflowModelSpec.scala (hunk 6 above)
Comment (austinzh) [reply mleap-tensorflow/src/test/scala/ml/combust/mleap/tensorflow/TensorflowModelSpec.scala:61]: FileUtil().rmRF(testFolder.toFile) added after we finished using it

**Comment by austinzh** [comment]
> looks good overall, one change is needed (i think) to make this backward compatible.

@ancasarb Since the `None` is cover the old format, It's OK to merge it?

**Review by ancasarb — APPROVED** [review]

**Comment by ancasarb** [comment]
> > looks good overall, one change is needed (i think) to make this backward compatible.
> 
> @ancasarb Since the `None` is cover the old format, It's OK to merge it?

@austinzh yes, sorry, I missed the `None` part initially, Ryan pointed it out earlier too.

File: mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowModel.scala [hunk 7]
```diff
@@ -64,26 +70,34 @@ case class TensorflowModel(@transient var graph: Option[tensorflow.Graph] = None
...
+  }
 
-    f(s)
+  private def getSessionS: (tensorflow.Session, tensorflow.Graph) = {
+    val dest = Files.createTempDirectory("saved_model")
+    val savedModelStream = new ZipInputStream(
+      new ByteArrayInputStream(modelBytes)
+    )
+    FileUtil().extract(savedModelStream, dest.toFile)
+    val modelBundle = tensorflow.SavedModelBundle.load(dest.toString)
```
[the remark this answers is not in the document]
Comment (giorgiomondauto) [reply mleap-tensorflow/src/main/scala/ml/combust/mleap/tensorflow/TensorflowModel.scala:98]: Hello. I have the TF2 model saved in a saved_model directory containing the saved_model.pb along with variables folder. When I load this model I am getting the following:

`Could not find SavedModel .pb or .pbtxt at supplied export directory path: /` . Any hint on how to solve this issue?

**Pull request merged**